chips/virtio/src/devices/virtio_gpu/mod.rs

```diff
@@ -0,0 +1,1069 @@
+            // Precompute the number of pixels in this draw area:
+            (width as usize)
+                .checked_mul(height as usize)
+                .ok_or(ErrorCode::INVAL)?,
+        ));
+
+        // Set the device state to busy and issue the callback in a deferred
+        // call:
+        self.state.set(VirtIOGPUState::SettingWriteFrame);
+        self.pending_deferred_call_mask
+            .set(deferred_call::PendingDeferredCall::SetWriteFrame);
+        self.deferred_call.set();
+
+        Ok(())
+    }
+
+    fn write(
+        &self,
+        buffer: SubSliceMut<'static, u8>,
+        continue_write: bool,
+    ) -> Result<(), ErrorCode> {
+        // Make sure we're idle:
+        let VirtIOGPUState::Idle = self.state.get() else {
+            return Err(ErrorCode::BUSY);
+        };
+
+        // If `continue_write` is false, we must reset `x_off` and
+        // `y_off`. Otherwise we start at the stored offset.
+        let (draw_rect, mut current_draw_offset, mut remaining_pixels) =
+            self.current_draw_area.get();
+        if !continue_write {
+            current_draw_offset = (0, 0);
+            // This multiplication must not overflow, as we've already performed
+            // it before in `set_write_area`:
+            remaining_pixels = (draw_rect.width as usize)
+                .checked_mul(draw_rect.height as usize)
+                .unwrap();
+        }
+        self.current_draw_area
+            .set((draw_rect, current_draw_offset, remaining_pixels));
+
+        // Ensure that this buffer is evenly divisible by PIXEL_STRIDE and that
+        // it can fit into the remaining part of the draw area:
+        if buffer.len() % PIXEL_STRIDE != 0 {
+            return Err(ErrorCode::INVAL);
+        }
+        if buffer.len() / PIXEL_STRIDE > remaining_pixels {
+            return Err(ErrorCode::SIZE);
+        }
+
+        // Now, the `TRANSFER_TO_HOST_2D` command can only copy rectangles.
+        // However, when we performed a partial write (let's say of just one
+        // pixel), then the current x offset will not perfectly line up with the
+        // left boundary of the overall draw rectangle. Similarly, when the
+        // buffer doesn't perfectly fill up the last row of pixels, we can't
+        // draw them together with the previous rows of the rectangle. Thus, a
+        // single `write` call may result in at most three underlying
+        // `TRANSFER_TO_HOST_2D` commands.
+        //
+        // We use a common subroutine to identify the next data to copy. We
+        // first store the overall subslice active range, and the offset in this
+        // subslice (0 right now!), and then let that subroutine handle the rest:
+        let write_buffer_subslice_range = buffer.active_range();
+        self.write_buffer_subslice_range.set((
+            write_buffer_subslice_range.start,
+            write_buffer_subslice_range.end,
+        ));
+        self.write_buffer_offset.set(0);
```

Comment:
I'm pretty confused with what's going on here. Particularly, what I expected you to do was to initially install a framebuffer of width*height u32s. Then on write you would modify the relevant pixels in the framebuffer, then flush the whole framebuffer (transfer to host 2d, followed by a resource flush).

Instead you've got a dummy framebuffer, then there's this whole comment about only copying rectangles that I don't follow at all. Can you explain what's going on? Was this an effort to save memory by not needing to represent the entire screen?